### PR TITLE
Variables in Nginx configuration

### PR DESCRIPTION
**Pull request by ShaneKilkelly**
## Description

This PR allows us to specify some parts of the Nginx configuration via environment variables. 
We achieve this by using [`envsubst`](https://github.com/a8m/envsubst) to fill in the variables in a template file, and write the resulting file to `/etc/nginx/nginx.conf`.

Users will be able to set `NGINX_WORKER_PROCESSES` and `NGINX_WORKER_CONNECTIONS`, to control the `worker_processes` and `worker_connections` variables, respectively.

## Changes
- Add `envsubst` binary, to the `vendor/` directory
- Remove the static `nginx.conf`
- Add a templates directory to the nginx confirguration (`/etc/nginx/templates`)
  - Add `nginx.conf.template`, replicating the content of the old, static config file
- Add an init-script (`init_scripts/01_nginx_config_template.sh`), which will run `envsubst` on the template, and write to `nginx.conf`
- In the template file, use the following environment variables: `NGINX_WORKER_PROCESSES` and `NGINX_WORKER_CONNECTIONS`, with appropriate fallback values.

## Rationale

- Nginx _does_ have limited support for variables, but only in version 1.19 onward. We are stuck with an older version, because we are stuck on an old version of the phusion base-image (see https://github.com/overleaf/issues/issues/3925#issuecomment-797439750)
  - Either way, this solution is more flexible, and under our control, which I think is a win
- `envsubst` is not available in the repos, for similar reasons

## Related issues / Pull Requests

- Fixes https://github.com/overleaf/issues/issues/3925

## Manual Testing

- [x] Tested with the toolkit, inspecting `/etc/nginx/nginx.conf` each time:
- [x] Set variables
- [x] Reset to defaults

## Contributor Agreement

- [x] I confirm I have signed the [Contributor License Agreement](https://github.com/overleaf/overleaf/blob/master/CONTRIBUTING.md#contributor-license-agreement)

### CODE REVIEW

**Comment by ShaneKilkelly** [comment]
Now that I'm thinking about it, maybe the variables should be:

- `NGINX_WORKER_PROCESSES` => `worker_processes`
- `NGINX_WORKER_CONNECTIONS` => `worker_connections`

?

File: nginx/nginx.conf.template [hunk 1]
```diff
@@ -1,10 +1,14 @@
+## ! ! ! ! ! ! ! ! ! ! ! ! ! ! ! ! ! ! ! ! ! ! ##
+## ! This file was generated from a template ! ##
+## ! See /etc/nginx/templates/               ! ##
+## ! ! ! ! ! ! ! ! ! ! ! ! ! ! ! ! ! ! ! ! ! ! ##
 daemon off;
 user www-data;
-worker_processes 4;
+worker_processes ${NGINX_WORKERS};
```
Comment (ShaneKilkelly) [inline nginx/nginx.conf.template:7]: Now that I'm thinking about it, maybe the variables should be:

- `NGINX_WORKER_PROCESSES` => `worker_processes`
- `NGINX_WORKER_CONNECTIONS` => `worker_connections`

?

**Review by mserranom — APPROVED** [review]
Looks good! Would just change the environment variables exactly as you've suggested.

**Comment by ShaneKilkelly** [comment]
@mserranom , thanks! I've fixed the variable names, and re-written the git history.

File: nginx/nginx.conf.template (hunk 1 above)
Comment (ShaneKilkelly) [reply nginx/nginx.conf.template]: Fixed

**Comment by ShaneKilkelly** [comment]
Just a reflection, but I guess we could do this with `m4`, if we didn't want to add a new dependency.
Having said that, `envsubst` is probably the right way to go.

**Comment by ShaneKilkelly** [comment]
To do after merge: update docs here https://github.com/overleaf/overleaf/wiki/HTTPS-reverse-proxy-using-Nginx

**Comment by ShaneKilkelly** [comment]
Requesting re-review on a new commit https://github.com/overleaf/overleaf/pull/853/commits/12e862c07844e979feb87afd46382f7269fbc576 , to address https://github.com/overleaf/issues/issues/3925#issuecomment-811777045

**Review by mserranom — APPROVED** [review]
This is great, thanks!

**Comment by ShaneKilkelly** [comment]
co-assigning @mserranom to merge for the release.

**Pull request merged**